File: Javascript/Algorithms/232.js
/**
 * Initialize your data structure here.
 */
var MyQueue = function() {
    this.storage = [];
    this.tmp = [];
};

/**
 * Push element x to the back of queue. 
 * @param {number} x
 * @return {void}
 *
 * Uses stack equivalent operation push.
 */
MyQueue.prototype.push = function(x) {
    this.storage.push(x);
};

/**
 * Removes the element from in front of queue and returns that element.
 * @return {number}
 *
 * First, transfer elements over from storage to tmp to bring them in FIFO
 * order. Then we use the stack equivalent operation pop.
 */
MyQueue.prototype.pop = function() {
    this._transfer();
    return this.tmp.pop();
};

/**
 * Get the front element.
 * @return {number}
 *
 * First, transfer elements over from storage to tmp to bring them in FIFO
 * order. Then we use the stack equivalent operation for peek/top.
 */
MyQueue.prototype.peek = function() {
    this._transfer();
    return this.tmp[this.tmp.length - 1];
};

/**
 * Returns whether the queue is empty.
 * @return {boolean}
 *
 * Uses operation equivalent for is_empty.
 */
MyQueue.prototype.empty = function() {
    return this.tmp.length == 0 && this.storage.length == 0;
};

/**
 * Transfers elements from storage to tmp, uses only stack equivalent
 * operations.
 */
MyQueue.prototype._transfer = function () {
    if (this.tmp.length == 0){
        while (this.storage.length) {
            this.tmp.push(this.storage.pop());
        }
    }
}

File: Rust/Algorithms/232.rs
struct MyQueue {
    storage: Vec<i32>,
    tmp: Vec<i32>,
}

impl MyQueue {
    /** Initialize your data structure here. */
    fn new() -> Self {
        Self {
            storage: vec![],
            tmp: vec![],
        }
    }

    /** Push element x to the back of queue.

    Uses Stack equivalent operation push (push).
    */
    fn push(&mut self, x: i32) {
        self.storage.push(x);
    }

    /** Removes the element from in front of queue and returns that element.

    Transfers elements from storage to tmp if necessary, this places elements in
    correct FIFO order. Uses only Stack equivalent operations.
    */
    fn pop(&mut self) -> i32 {
        self._transfer();
        self.tmp.pop().unwrap()
    }

    /** Get the front element.

    Transfers elements from storage to tmp if necessary, this places elements in
    correct FIFO order. Uses Stack equivalent operation last (peek, top).
    */
    fn peek(&mut self) -> i32 {
        self._transfer();
        *self.tmp.last().unwrap()
    }

    /** Returns whether the queue is empty.

    Uses stack equivalent operations len (is_empty)
    */
    fn empty(&self) -> bool {
        self.tmp.len() == 0 && self.storage.len() == 0
    }

    /** Transfer elements from storage to tmp. This uses stack equivalent
    operations len (is_empty), push (push), pop (pop).
    */
    fn _transfer(&mut self) {
        if self.tmp.len() == 0 {
            while self.storage.len() != 0 {
                self.tmp.push(self.storage.pop().unwrap());
            }
        }
    }
}
